### DISCUSSION TITLE
Detecting the end of a song

### DISCUSSION THREAD

**Discussion by vettorazi** [gilbarbara/react-spotify-web-playback#134] [Q&A] [answered]
Currently, I need to set a state in my app every time the song ends, the first logic I had was:
```
  callback={(state) => {
                    if (state.isPlaying) {
                      // PlayBlock(data, selectedBlock, setData);
                      console.log("spotify block is playing", state.progressMs);
                    }else{
                      console.log("spotify block is not playing");
                      if(state.progressMs>1000){
                        setSelectedBlock((prev) => prev + 1);
                      }
                    }
                  }}
```

But probably isn't a good idea. @gilbarbara or anyone, know a better way to accomplish this?
Thank you!

**Comment by gilbarbara** [comment, marked as answer]
Hey @vettorazi 
Take a look in the Type Definition for the [callback](https://github.com/gilbarbara/react-spotify-web-playback#props)

When the callback receives an event of type `track_update` it means the player just loaded a new track, so it finished playing the "previous" track.

**Comment by vettorazi** [comment]
Thanks @gilbarbara this worked!
I had to do this though:
```
              callback={(state) => {
                if(state.type==="player_update"&&!state.isPlaying){
                  console.log('spotify song ended')
                  setMusic(CurrentMusic + 1);
                }
              }}
```
